### ISSUE TITLE
BatchNormalization layer inside FrozenLayerWithBackprop NullPointerException

### ISSUE THREAD

**Issue by Rolnire** [deeplearning4j/deeplearning4j#9101]
#### Issue Description

When I try to use BatchNormalization layer inside FrozenLayerWithBackprop I am getting NullPointerException. I think it has something to do with mean/var cache. Same exception with GPU and CPU backend.

> WARNING: CuDNN BatchNormalization backprop execution failed - falling back on built-in implementation
java.lang.NullPointerException
at org.deeplearning4j.nn.layers.mkldnn.MKLDNNBatchNormHelper.backpropGradient(MKLDNNBatchNormHelper.java:101)
at org.deeplearning4j.nn.layers.normalization.BatchNormalization.backpropGradient(BatchNormalization.java:164)
at org.deeplearning4j.nn.layers.FrozenLayerWithBackprop.backpropGradient(FrozenLayerWithBackprop.java:62)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.calcBackpropGradients(MultiLayerNetwork.java:1946)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2761)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2704)
at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:170)
at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:63)
at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fitHelper(MultiLayerNetwork.java:2305)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:2263)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:2326)
at eu.extech.deep_learning.computer_vision.BatchNormTest.main(BatchNormTest.java:38)

> Exception in thread “main” java.lang.NullPointerException
at org.deeplearning4j.nn.layers.normalization.BatchNormalization.backpropGradient(BatchNormalization.java:296)
at org.deeplearning4j.nn.layers.FrozenLayerWithBackprop.backpropGradient(FrozenLayerWithBackprop.java:62)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.calcBackpropGradients(MultiLayerNetwork.java:1946)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2761)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2704)
at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:170)
at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:63)
at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fitHelper(MultiLayerNetwork.java:2305)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:2263)
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:2326)
at eu.extech.deep_learning.computer_vision.BatchNormTest.main(BatchNormTest.java:38)

This is my layer, how I am creating it:
> listBuilder.layer(new FrozenLayerWithBackprop(new BatchNormalization.Builder().eps(0.001).useLogStd(false).build()));

#### Version Information

* Deeplearning4j beta7 (I have some issues with downloading SNAPSHOT version now)
* Mac OS, Linux
* CUDA version 10.2
* NVIDIA driver version 440.33.01

**Comment by agibsonccc** [comment]
Interesting. Can you try this without mkldnn and see if it still causes this issue?
Your full environment here would be good. I see your version information, but I feel like a pom would be more appropriate here. I could only infer MKLDNN from your stack trace, so I want to make sure I'm not missing anything else.

**Comment by Rolnire** [comment]
For CPU backend this is my pom file:

	<artifactId>Dl4j-cpulibs</artifactId>
	<version>1.0.0-beta7</version>
	<packaging>pom</packaging>

	<name>Dl4j-cpulibs</name>
	<url>http://maven.apache.org</url>

	<properties>
		<nd4j.backend>nd4j-native-platform</nd4j.backend>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<shadedClassifier>bin</shadedClassifier>

		<java.version>1.8</java.version>
		<nd4j.version>1.0.0-beta7</nd4j.version>
		<dl4j.version>1.0.0-beta7</dl4j.version>
		<datavec.version>1.0.0-beta7</datavec.version>
		<arbiter.version>1.0.0-beta7</arbiter.version>
		<rl4j.version>1.0.0-beta7</rl4j.version>

		
		<dl4j.spark.version>1.0.0-beta7_spark_1</dl4j.spark.version>
		<datavec.spark.version>1.0.0-beta7_spark_1</datavec.spark.version>

		
		<scala.binary.version>2.11</scala.binary.version>

		<hadoop.version>2.2.0</hadoop.version>  
		<guava.version>19.0</guava.version>
		<logback.version>1.2.3</logback.version>
		<jfreechart.version>1.0.13</jfreechart.version>
		<jcommon.version>1.0.23</jcommon.version>
		<maven-compiler-plugin.version>3.6.1</maven-compiler-plugin.version>
		<maven-shade-plugin.version>2.4.3</maven-shade-plugin.version>
		<exec-maven-plugin.version>1.4.0</exec-maven-plugin.version>
		<maven.minimum.version>3.3.1</maven.minimum.version>
		<javafx.version>2.2.3</javafx.version>
		<aws.sdk.version>1.11.109</aws.sdk.version>
		<jackson.version>2.5.1</jackson.version>
		<scala.plugin.version>3.2.2</scala.plugin.version>

		<jersey.version>2.30.1</jersey.version>
		<jetty.version>9.4.27.v20200227</jetty.version>
	</properties>

	<repositories>
		<repository>
			<id>maven.restlet.org</id>
			<name>maven.restlet.org</name>
			<url>http://maven.restlet.org</url>
		</repository>

		<repository>
			<id>objectdb</id>
			<name>ObjectDB Repository</name>
			<url>https://m2.objectdb.com</url>
		</repository>

		<repository>
			<id>snapshots-repo</id>
			<url>https://oss.sonatype.org/content/repositories/snapshots</url>
			<releases>
				<enabled>false</enabled>
			</releases>
			<snapshots>
				<enabled>true</enabled>
				<updatePolicy>daily</updatePolicy>  
			</snapshots>
		</repository>
	</repositories>

	<parent>
		<groupId>org.deeplearning4j</groupId>
		<artifactId>deeplearning4j-parent</artifactId>
		<version>1.0.0-beta7</version>
	</parent>

	<dependencies>
		<dependency>
			<groupId>org.nd4j</groupId>
			<artifactId>${nd4j.backend}</artifactId>
			<version>${nd4j.version}</version>
			<exclusions>
				<exclusion>
					<groupId>org.bytedeco</groupId>
					<artifactId>leptonica-platform</artifactId>
				</exclusion>
				<exclusion>
					<groupId>ch.qos.logback</groupId>
					<artifactId>logback-classic</artifactId>
				</exclusion>
			</exclusions>
		</dependency> 

		<dependency>
			<groupId>org.deeplearning4j</groupId>
			<artifactId>deeplearning4j-nlp</artifactId>
			<version>${dl4j.version}</version>
		</dependency>

		<dependency>
			<groupId>org.deeplearning4j</groupId>
			<artifactId>deeplearning4j-zoo</artifactId>
			<version>${dl4j.version}</version>
		</dependency>

		
		<dependency>
			<groupId>com.workday</groupId>
			<artifactId>timeseries-forecast</artifactId>
			<version>1.1.1</version>
		</dependency>

		<dependency>
			<groupId>org.eclipse.jetty</groupId>
			<artifactId>jetty-server</artifactId>
			<version>${jetty.version}</version>
		</dependency>

		<dependency>
			<groupId>org.eclipse.jetty</groupId>
			<artifactId>jetty-servlet</artifactId>
			<version>${jetty.version}</version>
		</dependency>

		<dependency>
			<groupId>org.glassfish.jersey.core</groupId>
			<artifactId>jersey-server</artifactId>
			<version>${jersey.version}</version>
		</dependency>

		<dependency>
			<groupId>org.glassfish.jersey.media</groupId>
			<artifactId>jersey-media-json-jettison</artifactId>
			<version>${jersey.version}</version>
		</dependency>

		<dependency>
			<groupId>org.glassfish.jersey.containers</groupId>
			<artifactId>jersey-container-servlet-core</artifactId>
			<version>${jersey.version}</version>
		</dependency>

		<dependency>
			<groupId>org.glassfish.jersey.inject</groupId>
			<artifactId>jersey-hk2</artifactId>
			<version>${jersey.version}</version>
		</dependency>

		<dependency>
			<groupId>com.objectdb</groupId>
			<artifactId>objectdb</artifactId>
			<version>2.8.2</version>
		</dependency>

		<dependency>
			<groupId>org.eclipse.persistence</groupId>
			<artifactId>javax.persistence</artifactId>
			<version>2.1.0</version>
		</dependency>

		<dependency>
			<groupId>javax.transaction</groupId>
			<artifactId>jta</artifactId>
			<version>1.1</version>
		</dependency>

		<dependency>
			<groupId>org.deeplearning4j</groupId>
			<artifactId>deeplearning4j-ui</artifactId>
			<version>${dl4j.version}</version>
		</dependency>
	</dependencies>

	<build>
		<plugins>
			<plugin>
				<artifactId>maven-enforcer-plugin</artifactId>
				<version>1.0.1</version>
				<executions>
					<execution>
						<id>enforce-default</id>
						<goals>
							<goal>enforce</goal>
						</goals>
						<configuration>
							<rules>
								<requireMavenVersion>
									<version>[${maven.minimum.version},)</version>
									<message>********** Minimum Maven Version is
										${maven.minimum.version}. Please upgrade Maven before
										continuing (run "mvn --version" to check). **********</message>
								</requireMavenVersion>
							</rules>
						</configuration>
					</execution>
				</executions>
			</plugin>

			
			<plugin>
				<groupId>net.revelc.code.formatter</groupId>
				<artifactId>formatter-maven-plugin</artifactId>
				<version>2.0.0</version>
				<configuration>
					
					<directories>
						<directory>dl4j-examples</directory>
						<directory>dl4j-spark-examples</directory>
						<directory>datavec-examples</directory>
						<directory>dl4j-cuda-specific-examples</directory>
						<directory>nd4j-examples</directory>
						<directory>rl4j-examples</directory>
						<directory>arbiter-examples</directory>
						<directory>lstm-hdfs</directory>
					</directories>
				</configuration>
			</plugin>

			<plugin>
				<groupId>com.lewisd</groupId>
				<artifactId>lint-maven-plugin</artifactId>
				<version>0.0.11</version>
				<configuration>
					<failOnViolation>true</failOnViolation>
					<onlyRunRules>
						<rule>DuplicateDep</rule>
						<rule>RedundantPluginVersion</rule>
						<rule>VersionProp</rule>
						<rule>DotVersionProperty</rule>
					</onlyRunRules>
				</configuration>
				<executions>
					<execution>
						<id>pom-lint</id>
						<phase>validate</phase>
						<goals>
							<goal>check</goal>
						</goals>
					</execution>
				</executions>
			</plugin>

			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-compiler-plugin</artifactId>
				<configuration>
					<source>${java.version}</source>
					<target>${java.version}</target>
				</configuration>
			</plugin>
		</plugins>

		<pluginManagement>
			<plugins>
				<plugin>
					<groupId>org.eclipse.m2e</groupId>
					<artifactId>lifecycle-mapping</artifactId>
					<version>1.0.0</version>
					<configuration>
						<lifecycleMappingMetadata>
							<pluginExecutions>
								<pluginExecution>
									<pluginExecutionFilter>
										<groupId>com.lewisd</groupId>
										<artifactId>lint-maven-plugin</artifactId>
										<versionRange>[0.0.11,)</versionRange>
										<goals>
											<goal>check</goal>
										</goals>
									</pluginExecutionFilter>
									<action>
										<ignore />
									</action>
								</pluginExecution>
							</pluginExecutions>
						</lifecycleMappingMetadata>
					</configuration>
				</plugin>
			</plugins>
		</pluginManagement>
	</build>

How can I easily disable MKLDNN?

**Comment by agibsonccc** [comment]
You can do that here: https://github.com/eclipse/deeplearning4j/blob/2a488efb1b723386eda004d8b35371a1f9e4d46b/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/BatchNormalization.java#L467

**Comment by Rolnire** [comment]
Sorry for really late response, but I finally had time to try it. I set helperAllowFallback to false, but exception is still there:

	Exception` in thread "main" java.lang.RuntimeException: Error during BatchNormalization CuDNN helper backprop - 	isCudnnAllowFallback() is set to false

	at org.deeplearning4j.nn.layers.normalization.BatchNormalization.backpropGradient(BatchNormalization.java:178)
	at org.deeplearning4j.nn.layers.FrozenLayerWithBackprop.backpropGradient(FrozenLayerWithBackprop.java:62)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.calcBackpropGradients(MultiLayerNetwork.java:1946)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2761)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2704)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:170)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:63)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fitHelper(MultiLayerNetwork.java:2305)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:2263)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:2326)
	at eu.extech.deep_learning.computer_vision.BatchNormTest.main(BatchNormTest.java:46)
	Caused by: java.lang.NullPointerException: Cannot invoke "org.nd4j.linalg.api.ndarray.INDArray.dataType()" because "this.meanCache" is null
	at org.deeplearning4j.nn.layers.mkldnn.MKLDNNBatchNormHelper.backpropGradient(MKLDNNBatchNormHelper.java:101)
	at org.deeplearning4j.nn.layers.normalization.BatchNormalization.backpropGradient(BatchNormalization.java:164)
	... 11 more

**Comment by agibsonccc** [comment]
Closing — this issue is over 2 years old with no recent activity or reproduction steps. The underlying code has been significantly rewritten. If you still experience this issue with a recent version, please open a new issue.
